### PR TITLE
cmake(macOS): Remove hard-coded Homebrew dependencies path

### PR DESCRIPTION
**Pull request by lucydodo**
Related: 7748dae8eb0f45f9bc34ecddfc80683c35e3f59a

I reorganized the `CMakeLists.txt` file based on the discussion in the commit comment above.
However, I used macros instead of functions becaused of the scope issue.

Overall, it looks good to me, but the execution time of the initial cmake command was increased by running the `brew` command. Of course, generally don't use usally configure build very often, so it's not a big deal. (See *1)
Alternatively, I could change it to only get the 'Homebrew' path and then manually inject the package name afterward:
This will change the `brew` command to a maximum of 1 instead of the existing maximum of 6.

@scottfurry I would be greatful if you could review this PR. Thanks.

> *1: Benchmark results before and after improvements.

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `cmake -DcustomTap=1 -Dsqlcipher=1 ../before` | 611.7 ± 31.2 | 577.4 | 666.6 | 1.00 |
| `cmake -DcustomTap=1 -Dsqlcipher=1 ../after`    | 5541.7 ± 143.7 | 5370.9 | 5769.9 | 9.06 ± 0.52 |

### CODE REVIEW

**Comment by scottfurry** [comment]
@lucydodo - that looks close to what we discussed elsewhere.

I'm leery of macros. Scope and Variables can get messy in these context if one is not careful. However, if it works and there isn't a _better_ suggestion then go with it.

One idea to consider is the idea of [CMake Modules](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#modules). Modules are used in the CMake ecosystem to integrate specific software packages into the CMake build generation process. (e.g. look for FindBoost.cmake - POSIX: /usr/share/cmake/Modules/FindBoost.cmake to aid CMake pulling in Boost libraries).

Ideally, KitWare would have a `FindHomeBrew.cmake` already but it seems were treading on virgin territory. There's a _feature request_ that could be made to CMake devs.

The idea of an external module adding to the build target seems more in line with current CMake practices.

**Comment by lucydodo** [comment]
Thank you for your review.
I hand't thought of the 'module', but it sound resonable. I'll look into it and se if we can apply it.

**Comment by justinclift** [comment]
> Ideally, KitWare would have a FindHomeBrew.cmake already but it seems were treading on virgin territory. There's a feature request that could be made to CMake devs.

That's a good idea.  Shouldn't hurt to at least ask. :smile:

@scottfurry @lucydodo Either of you up for doing that asking?

**Comment by lucydodo** [comment]
@scottfurry Okay, so now I've moved code into a module.
However, I'm not sure if my approach is best practice.

Here are some new benchmark results:
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `current` | 6.611 ± 0.281 | 6.317 | 7.166 | 4.51 ± 0.87 |
| `before`  | 1.466 ± 0.277 | 1.046 | 1.842 | 1.00 |
| `after`   | 1.688 ± 0.479 | 1.117 | 2.634 | 1.15 ± 0.39 |

> `current`: Relative to the `master` branch of the current code-base
> `before`: When the PR was made (when the 'macro' was used)
> `after`: At the current time (when using the 'module')

**Comment by lucydodo** [comment]
@justinclift I don't know much about the CMake eco-system, so this is going to be difficult for me. Sorry for that :(

**Comment by scottfurry** [comment]
@lucydodo - execution time is <10s. Or to put it another way, are we absolutely beholding to shaving milliseconds here?
Does code _do the job_?
Does code _use as few resources of the system as possible_?
Is code _understandable and/or make sense_ when reviewed a day or so later?

So long as we are not causing a system to suddenly release _manufacturer's installed smoke_, or otherwise emulate a funeral pyre, I think we're good. I have enough gray hairs for the group. Don't stress yourself out over this. And any gray hairs you do develop you would still lose - even if I just used my chest hairs alone for the comparison. :stuck_out_tongue_winking_eye: 

_deep cleans breath_
_and relax_

**Comment by scottfurry** [comment]
> > Ideally, KitWare would have a FindHomeBrew.cmake already but it seems were treading on virgin territory. There's a feature request that could be made to CMake devs.
> 
> That's a good idea. Shouldn't hurt to at least ask. 😄
> 
> @scottfurry @lucydodo Either of you up for doing that asking?

I would open my mouth...
I had a look on [CMake's Issue Board](https://gitlab.kitware.com/cmake/cmake/-/issues), a gitlab instance BTW, and there a few open bugs reporting issues with libraries installed via homebrew. The quick scan I did indicated something I had already encountered, homebrew's relentless pursuit of latest/greatest releases braking things.

I think we should hammer down a few things on our own for the moment.

@lucydodo - [Modern CMake](https://www.amazon.com/Modern-CMake-Discover-approach-packaging/dp/1801070059/) - maybe find a copy. There is also the [Awesome CMake](https://github.com/onqtam/awesome-cmake) repository here on GitHub.

**Comment by justinclift** [comment]
> I think we should hammer down a few things on our own for the moment.

No worries, we can do it later on at some point if we feel like it. :smile:

**Comment by lucydodo** [comment]
@scottfurry Right, it's not a big deal. I was just worried that someone might be inconveneienced by my changes :)
Anyway, does this code need a little more tweaking? or would it be okay to merge it?

**Comment by lucydodo** [comment]
Fortunately, I found a few issues before the merge, and I'll reopen it once they're fixed.

**Comment by lucydodo** [comment]
There is a problem with the `customTap` option not finding the correct path when used on Intel architecture Macs,
so I'm working on that. I'll try to fix it this week at the latest. 😄

**Comment by justinclift** [comment]
No worries, it's not urgent.  Was just checking things as my focus has been elsewhere for a bit. :smile:

**Comment by lucydodo** [comment]
Now we're ready to merge :)

File: CMakeLists.txt [hunk 1]
```diff
@@ -108,28 +108,11 @@ endif()
...
-        list(PREPEND CMAKE_PREFIX_PATH "/opt/homebrew/opt/db4subqt@5")
-        list(PREPEND CMAKE_PREFIX_PATH "/opt/homebrew/opt/db4subsqlitefts@5")
-
-        if(sqlcipher)
-            list(APPEND SQLCIPHER_INCLUDE_DIR "/opt/homebrew/include")
-            list(APPEND SQLCIPHER_LIBRARY "/opt/homebrew/opt/db4subsqlcipher/lib/libsqlcipher.0.dylib")
-        endif()
+    find_package(Homebrew REQUIRED)
+    if (customTap AND sqlcipher)
+        list(PREPEND CMAKE_PREFIX_PATH ${Qt5_DIR} ${SQLCipher_INCLUDE_DIR} ${SQLCipher_INCLUDE_DIR} ${SQLite_DIR})
```
Comment (justinclift) [inline CMakeLists.txt:113]: `${SQLCipher_INCLUDE_DIR}` seems to be included twice here.  That doesn't *seem* correct to me, but I could be wrong. :smile:

Comment (lucydodo) [reply CMakeLists.txt]: Oh no, that's a mistake. Thanks for letting me know :). Resolve via [`29838cb` (#3472)](https://github.com/sqlitebrowser/sqlitebrowser/pull/3472/commits/29838cbacb6d98e50dad9ab4234397561e906577)

**Comment by lucydodo** [comment]
Once I'm done with some pressing issues on the project, I'll be writing a new wiki on how to build our project, and along with that I'll be makign some small changes to the mechanism for finding build dependency paths.
So I'm going to close this PR for now. :)

**Comment by justinclift** [comment]
Should we delete this branch, or is it better to leave hanging around for a while? :smile:

**Comment by lucydodo** [comment]
Good catch. I just deleted it. :)

**Pull request closed without merging**